### PR TITLE
FOGL-8323: addition of join in Postgres and sqlitelb storage plugins

### PR DESCRIPTION
**Pull request by pintomax**
FOGL-8323: Initial addition of join in retrieve

### CODE REVIEW

**Comment by pintomax** [comment]
Similar fixes are required in "sqlitelb" storage plugin

File: C/plugins/storage/postgres/connection.cpp [hunk 1]
```diff
@@ -3572,14 +3678,21 @@ bool Connection::appendTables(const Value& document, SQLBuffer& sql, int level)
 				raiseError("commonRetrieve", "Joining table name is not a string");
 				return false;
 			}
-			sql.append(", fledge.");
-			sql.append(name.GetString());
-			sql.append(" ");
-			sql.append(tag);
+			//sql.append(", fledge.");
```
Comment (MarkRiddoch) [inline C/plugins/storage/postgres/connection.cpp:3681]: Let's delete the commented out code

File: C/plugins/storage/postgres/plugin.cpp [hunk 2]
```diff
@@ -122,7 +122,8 @@ ConnectionManager *manager = (ConnectionManager *)handle;
 Connection        *connection = manager->allocate();
 std::string results;
 
-	bool rval = connection->retrieve(std::string(OR_DEFAULT_SCHEMA(schema)) + "." + std::string(table), std::string(query), results);
+	//bool rval = connection->retrieve(std::string(OR_DEFAULT_SCHEMA(schema)) + "." + std::string(table), std::string(query), results);
```
Comment (MarkRiddoch) [inline C/plugins/storage/postgres/plugin.cpp:125]: Please delete commented out code

**Review by MarkRiddoch — CHANGES_REQUESTED** [review]
Looks reasonable, but please delete the commented out code

**Review by ashish-jabble — CHANGES_REQUESTED** [review]
This is not working with the mentioned problem given in original JIRA.

Or something is pending on bucket service to adapt the PostgreSQL Join? 

It would be great if we add **unit tests** and this will also help to find the regression if any with the change.

**Comment by pintomax** [comment]
Steps to add in fledge/foglamp setup

(1)
# sudo -u postgres psql -d fledge

CREATE SCHEMA Bucket;
GRANT USAGE ON SCHEMA Bucket TO PUBLIC;

CREATE DATABASE Bucket WITH
    ENCODING = 'UTF8';
GRANT CONNECT ON DATABASE Bucket TO PUBLIC;

CREATE TABLE bucket.buckets (
    filename character varying(255),
    id integer NOT NULL
);

ALTER TABLE ONLY bucket.buckets
    ADD CONSTRAINT buckets_pkey PRIMARY KEY (id);

CREATE TABLE bucket.bucket_attributes (
    value character varying(80),
    name character varying(80),
    id integer
);

INSERT INTO bucket.bucket_attributes VALUES ('intel', 'architecture', 0);
INSERT INTO bucket.bucket_attributes VALUES ('Model for ADM filter prediction', 'description', 0);
INSERT INTO bucket.bucket_attributes VALUES ('gpu', 'hardware', 0);
INSERT INTO bucket.bucket_attributes VALUES ('ADM', 'name', 0);
INSERT INTO bucket.bucket_attributes VALUES ('model', 'type', 0);
INSERT INTO bucket.bucket_attributes VALUES ('1.0', 'version', 0);
INSERT INTO bucket.bucket_attributes VALUES ('Testing', 'description', 1);
INSERT INTO bucket.bucket_attributes VALUES ('Test', 'name', 1);
INSERT INTO bucket.bucket_attributes VALUES ('core', 'type', 1);
INSERT INTO bucket.bucket_attributes VALUES ('1.0', 'version', 1);

INSERT INTO bucket.buckets VALUES ('/usr/local/foglamp/data/buckets/0/0', 0);
INSERT INTO bucket.buckets VALUES ('/usr/local/foglamp/data/buckets/0/1', 1);


If needed set privileges for root or fledge user

ALTER TABLE bucket.bucket_attributes OWNER TO root;
ALTER TABLE bucket.buckets OWNER TO root;

(2)
start flege and get storage service service_port
curl 127.0.0.1:$core_port/fledge/service

(3)
Storage query

curl -X PUT 127.0.0.1:$storage_service_port/storage/schema/Bucket/table/buckets/query -d '
{ "where" : { "column" :  "id", "condition" : "!=", "value" : "-1" },  "join" : {"table" :{ "name" : "bucket_attributes", "column" : "id"}, "on" : "id", "query" : { "where" : { "column" : "name", "condition" : "=", "value" : "type", "and" : { "column" : "value", "condition" : "=", "value" : "bla" } } } }, "return" : [ "id" ]}' 

==> 0 rows

curl -X PUT 127.0.0.1:$storage_service_port/storage/schema/Bucket/table/buckets/query -d '
{ "where" : { "column" :  "id", "condition" : "!=", "value" : "-1" },  "join" : {"table" :{ "name" : "bucket_attributes", "column" : "id"}, "on" : "id", "query" : { "where" : { "column" : "name", "condition" : "=", "value" : "type", "and" : { "column" : "value", "condition" : "=", "value" : "model" } } } }, "return" : [ "id" ]}' 

==> 1 row

**Comment by ashish-jabble** [comment]
**On PostgreSQL**

I see on adding a bucket now error logs are coming though bucket has been created.

```
curl -s -F bucket=@data/bucket/adm_ld_ae.tflite -F attributes='{"name": "ADM", "architecture": "intel", "hardware": "gpu", "version": "1.0","description": "Model for ADM filter prediction", "type": "model"}' POST localhost:34489/bucket
```

**Syslogs**

```
aj-ThinkPad-E450 FogLAMP Storage[10848]: ERROR: PostgreSQL storage plugin raising error: ERROR:  column reference "value" is ambiguous#012LINE 1: ... '-1' AND t0.id = t1.id AND  t1.name = 'name' AND "value" = ...#012                                                             ^
aj-ThinkPad-E450 FogLAMP BucketServer[11385]: ERROR: Query table: ERROR:  column reference "value" is ambiguousLINE 1: ... '-1' AND t0.id = t1.id AND  t1.name = 'name' AND "value" = ...  
```

And  for matching an attribute looks good.

**Comment by ashish-jabble** [comment]
**On SQLitelb** - REGRESSION

A) adding a bucket service

```
Dec 14 12:21:43 aj-ThinkPad-E450 FogLAMP[13513] INFO: service_registry: foglamp.services.core.service_registry.service_registry: Registered service instance id=c120ee9e-b3fb-44b2-8321-4151504d3726: <BucketServer, type=BucketStorage, protocol=http, address=localhost, service port=45615, management port=38477, status=1>



Dec 14 12:21:44 aj-ThinkPad-E450 FogLAMP Storage[13604]: ERROR: SQLite 3 lb storage plugin raising error: no such table: foglamp.buckets
Dec 14 12:21:44 aj-ThinkPad-E450 FogLAMP Storage[13604]: ERROR: SQL statement: SELECT max("id") AS "max_id" FROM foglamp.buckets WHERE id > '0';
Dec 14 12:21:44 aj-ThinkPad-E450 FogLAMP BucketServer[13870]: ERROR: Query table: no such table: foglamp.buckets
```
B) On adding a bucket - it says with 200 but it totally breaks. See error logs 

```
Dec 14 12:21:44 aj-ThinkPad-E450 FogLAMP BucketServer[13870]: ERROR: Query table: no such table: foglamp.buckets
Dec 14 12:22:57 aj-ThinkPad-E450 FogLAMP Storage[13604]: ERROR: SQLite 3 lb storage plugin raising error: ambiguous column name: value
Dec 14 12:22:57 aj-ThinkPad-E450 FogLAMP Storage[13604]: ERROR: SQL statement: SELECT t0.id, * FROM Bucket.buckets t0, Bucket.bucket_attributes t1, Bucket.bucket_attributes t2, Bucket.bucket_attributes t3, Bucket.bucket_attributes t4, Bucket.bucket_attributes t5, Bucket.bucket_attributes t6 WHERE t0.id != '-1' AND t0.id = t1.id AND  t1.name = 'name' AND value = 'ADM' AND t1.id = t2.id  AND t2.name = 'architecture' AND value = 'intel' AND t2.id = t3.id  AND t3.name = 'hardware' AND value = 'gpu' AND t3.id = t4.id  AND t4.name = 'version' AND value = '1.0' AND t4.id = t5.id  AND t5.name = 'description' AND value = 'Model for ADM filter prediction' AND t5.id = t6.id  AND t6.name = 'type' AND value = 'model';
Dec 14 12:22:57 aj-ThinkPad-E450 FogLAMP BucketServer[13870]: ERROR: Query table: ambiguous column name: value
Dec 14 12:22:57 aj-ThinkPad-E450 FogLAMP Storage[13604]: ERROR: SQLite 3 lb storage plugin raising error: no such table: foglamp.buckets
Dec 14 12:22:57 aj-ThinkPad-E450 FogLAMP Storage[13604]: ERROR: SQL statement: SELECT  * FROM foglamp.buckets WHERE id = '0';
Dec 14 12:22:57 aj-ThinkPad-E450 FogLAMP BucketServer[13870]: ERROR: Query table: no such table: foglamp.buckets
Dec 14 12:22:58 aj-ThinkPad-E450 FogLAMP Storage[13604]: ERROR: SQLite 3 lb storage plugin raising error: no such table: foglamp.bucket_attributes
Dec 14 12:22:58 aj-ThinkPad-E450 FogLAMP Storage[13604]: ERROR: SQL statement: SELECT  * FROM foglamp.bucket_attributes WHERE id = '0' AND name = 'architecture';
Dec 14 12:22:58 aj-ThinkPad-E450 FogLAMP BucketServer[13870]: ERROR: Query table: no such table: foglamp.bucket_attributes
Dec 14 12:22:58 aj-ThinkPad-E450 FogLAMP Storage[13604]: ERROR: SQLite 3 lb storage plugin raising error: no such table: foglamp.bucket_attributes
Dec 14 12:22:58 aj-ThinkPad-E450 FogLAMP Storage[13604]: ERROR: SQL statement: SELECT  * FROM foglamp.bucket_attributes WHERE id = '0' AND name = 'description';
Dec 14 12:22:58 aj-ThinkPad-E450 FogLAMP BucketServer[13870]: ERROR: Query table: no such table: foglamp.bucket_attributes
Dec 14 12:22:58 aj-ThinkPad-E450 FogLAMP Storage[13604]: ERROR: SQLite 3 lb storage plugin raising error: no such table: foglamp.bucket_attributes
Dec 14 12:22:58 aj-ThinkPad-E450 FogLAMP Storage[13604]: ERROR: SQL statement: SELECT  * FROM foglamp.bucket_attributes WHERE id = '0' AND name = 'hardware';
Dec 14 12:22:58 aj-ThinkPad-E450 FogLAMP BucketServer[13870]: ERROR: Query table: no such table: foglamp.bucket_attributes
Dec 14 12:22:58 aj-ThinkPad-E450 FogLAMP Storage[13604]: ERROR: SQLite 3 lb storage plugin raising error: no such table: foglamp.bucket_attributes
Dec 14 12:22:58 aj-ThinkPad-E450 FogLAMP Storage[13604]: ERROR: SQL statement: SELECT  * FROM foglamp.bucket_attributes WHERE id = '0' AND name = 'name';
Dec 14 12:22:58 aj-ThinkPad-E450 FogLAMP BucketServer[13870]: ERROR: Query table: no such table: foglamp.bucket_attributes
Dec 14 12:22:58 aj-ThinkPad-E450 FogLAMP Storage[13604]: ERROR: SQLite 3 lb storage plugin raising error: no such table: foglamp.bucket_attributes
Dec 14 12:22:58 aj-ThinkPad-E450 FogLAMP Storage[13604]: ERROR: SQL statement: SELECT  * FROM foglamp.bucket_attributes WHERE id = '0' AND name = 'type';
Dec 14 12:22:58 aj-ThinkPad-E450 FogLAMP BucketServer[13870]: ERROR: Query table: no such table: foglamp.bucket_attributes
Dec 14 12:22:58 aj-ThinkPad-E450 FogLAMP Storage[13604]: ERROR: SQLite 3 lb storage plugin raising error: no such table: foglamp.bucket_attributes
Dec 14 12:22:58 aj-ThinkPad-E450 FogLAMP Storage[13604]: ERROR: SQL statement: SELECT  * FROM foglamp.bucket_attributes WHERE id = '0' AND name = 'version';
Dec 14 12:22:58 aj-ThinkPad-E450 FogLAMP BucketServer[13870]: ERROR: Query table: no such table: foglamp.bucket_attributes
```

**Review by MarkRiddoch — APPROVED** [review]

**Review by ashish-jabble — APPROVED** [review]
It would be nice to have unit tests. Otherwise good.

**Pull request merged**